src/mlpack/core/tree/vantage_point_tree/dual_tree_traverser_impl.hpp

```diff
@@ -0,0 +1,567 @@
+    }
+
+    // Restore the main traversal information.
+    rule.TraversalInfo() = traversalInfo;
+
+    // Now recurse down the right query node.
+    leftScore = rule.Score(*queryNode.Right(), *referenceNode.Left());
+    leftInfo = rule.TraversalInfo();
+    rule.TraversalInfo() = traversalInfo;
+    rightScore = rule.Score(*queryNode.Right(), *referenceNode.Right());
+    numScores += 2;
+
+    if (leftScore < rightScore)
+    {
+      // Recurse to the left.  Restore the left traversal info.  Store the right
+      // traversal info.
+      rightInfo = rule.TraversalInfo();
+      rule.TraversalInfo() = leftInfo;
+      Traverse(*queryNode.Right(), *referenceNode.Left());
+
+      // Is it still valid to recurse to the right?
+      rightScore = rule.Rescore(*queryNode.Right(), *referenceNode.Right(),
+          rightScore);
+
+      if (rightScore != DBL_MAX)
+      {
+        // Restore the right traversal info.
+        rule.TraversalInfo() = rightInfo;
+        Traverse(*queryNode.Right(), *referenceNode.Right());
+      }
+      else
+        ++numPrunes;
+    }
+    else if (rightScore < leftScore)
+    {
+      // Recurse to the right.
+      Traverse(*queryNode.Right(), *referenceNode.Right());
+
+      // Is it still valid to recurse to the left?
+      leftScore = rule.Rescore(*queryNode.Right(), *referenceNode.Left(),
+          leftScore);
+
+      if (leftScore != DBL_MAX)
+      {
+        // Restore the left traversal info.
+        rule.TraversalInfo() = leftInfo;
+        Traverse(*queryNode.Right(), *referenceNode.Left());
+      }
+      else
+        ++numPrunes;
+    }
+    else
+    {
+      if (leftScore == DBL_MAX)
+      {
+        numPrunes += 2;
+      }
+      else
+      {
+        // Choose the left first.  Restore the left traversal info.  Store the
+        // right traversal info.
+        rightInfo = rule.TraversalInfo();
+        rule.TraversalInfo() = leftInfo;
+        Traverse(*queryNode.Right(), *referenceNode.Left());
+
+        // Is it still valid to recurse to the right?
+        rightScore = rule.Rescore(*queryNode.Right(), *referenceNode.Right(),
+            rightScore);
+
+        if (rightScore != DBL_MAX)
+        {
+          // Restore the right traversal info.
+          rule.TraversalInfo() = rightInfo;
+          Traverse(*queryNode.Right(), *referenceNode.Right());
+        }
+        else
+          ++numPrunes;
+      }
+    }
+  }
+}
+
+template<typename MetricType,
+         typename StatisticType,
+         typename MatType,
+         template<typename BoundMetricType, typename...> class BoundType,
+         template<typename SplitBoundType, typename SplitMatType>
+             class SplitType>
+template<typename RuleType>
+void VantagePointTree<MetricType, StatisticType, MatType, BoundType, SplitType>::
+DualTreeTraverser<RuleType>::Traverse(
+    const size_t queryIndex,
+    VantagePointTree<MetricType, StatisticType, MatType, BoundType, SplitType>&
+        referenceNode)
+{
```

Comment:
Would it be possible to use the `SingleTreeTraverser` here, or possibly to merge both the `DualTreeTraverser` and `SingleTreeTraverser` into a single class?